lib/cn/csched_sp3.c

```diff
@@ -1593,81 +1598,209 @@ sp3_comp_thread_name(
+        goto out;
+    }
 
-        jps->hdr = true;
-        bufsz -= n;
-        buf += n;
+    if (ev(!cJSON_AddStringToObject(node, "rule", cn_rule2str(w->cw_rule)))) {
+        err = merr(ENOMEM);
+        goto out;
     }
 
-    tm = (jclock_ns - w->cw_t0_enqueue) / NSEC_PER_SEC;
-    snprintf(tmbuf, sizeof(tmbuf), "%lu:%02lu", (tm / 60) % 60, tm % 60);
+    if (ev(!cJSON_AddNumberToObject(node, "qnum", w->cw_qnum))) {
+        err = merr(ENOMEM);
+        goto out;
+    }
+
+    if (ev(!cJSON_AddNumberToObject(node, "busy", atomic_read(&w->cw_node->tn_busycnt) >> 16))) {
+        err = merr(ENOMEM);
+        goto out;
+    }
+
+    if (ev(!cJSON_AddNumberToObject(node, "kvsets", w->cw_kvset_cnt))) {
+        err = merr(ENOMEM);
+        goto out;
+    }
+
+    if (ev(!cJSON_AddNumberToObject(node, "allocated_length_mb",
+            cn_ns_alen(&w->cw_ns) >> MB_SHIFT))) {
+        err = merr(ENOMEM);
+        goto out;
+    }
+
+    if (ev(!cJSON_AddNumberToObject(node, "compacted_length_mb",
+            cn_ns_clen(&w->cw_ns) >> MB_SHIFT))) {
+        err = merr(ENOMEM);
+        goto out;
+    }
+
+    if (ev(!cJSON_AddNumberToObject(node, "pcap", w->cw_ns.ns_pcap))) {
+        err = merr(ENOMEM);
+        goto out;
+    }
+
+    if (ev(!cJSON_AddNumberToObject(node, "compc", w->cw_compc))) {
+        err = merr(ENOMEM);
+        goto out;
+    }
+
+    if (ev(!cJSON_AddNumberToObject(node, "dgen", w->cw_dgen_hi_min))) {
+        err = merr(ENOMEM);
+        goto out;
+    }
+
+    if (ev(!cJSON_AddNumberToObject(node, "hblocks", w->cw_nh))) {
+        err = merr(ENOMEM);
+        goto out;
+    }
+
+    if (ev(!cJSON_AddNumberToObject(node, "kblocks", w->cw_nk))) {
+        err = merr(ENOMEM);
+        goto out;
+    }
+
+    if (ev(!cJSON_AddNumberToObject(node, "vblocks", w->cw_nv))) {
+        err = merr(ENOMEM);
+        goto out;
+    }
+
+    if (ev(!cJSON_AddNumberToObject(node, "root_allocated_length_mb",
+            w->cw_est.cwe_samp.r_alen >> MB_SHIFT))) {
+        err = merr(ENOMEM);
+        goto out;
+    }
+
+    if (ev(!cJSON_AddNumberToObject(node, "leaf_allocated_length_mb",
+            w->cw_est.cwe_samp.l_alen >> MB_SHIFT))) {
+        err = merr(ENOMEM);
+        goto out;
+    }
+
+    if (ev(!cJSON_AddNumberToObject(node, "leaf_compacted_length_mb",
+            w->cw_est.cwe_samp.l_good >> MB_SHIFT))) {
+        err = merr(ENOMEM);
+        goto out;
+    }
+
+    if (ev(!cJSON_AddStringToObject(node, "wmesg", sts_job_wmesg_get(job)))) {
+        err = merr(ENOMEM);
+        goto out;
+    }
+
+    if (ev(!cJSON_AddStringToObject(node, "time", tmbuf))) {
+        err = merr(ENOMEM);
+        goto out;
+    }
+
+    if (ev(!cJSON_AddStringToObject(node, "thread_name", w->cw_threadname))) {
+        err = merr(ENOMEM);
+        goto out;
+    }
+
+    if (ev(!cJSON_AddItemToArray(jobs, node))) {
+        err = merr(ENOMEM);
+        goto out;
+    }
+
+out:
+    if (err)
+        cJSON_Delete(node);
```

Comment:
Your idea about the bitwise is nice. Will look into it.

Reply:
Yeah, I use that in contexts where errors are extremely unlikely, can really eliminate a lot of mindless boilerplate.   Some people would write a horrid macro to do it, leveraging local vars, ugh, I wouldn't go there!  Just so long as you can keep calling into cjson having had a failure, looks like there is no issue based on my casual look at it...

Reply:
Or one giant if expression since they share the error value:
```
if (!cJSON_this() ||
    !cJSON_that() ||
    !cJSON_blahblah)
{
    status = ;
    goto end;
}
```

Reply:
Already went with Greg's bitwise idea. About to push updates.